File: app/src/main/java/za/co/xisystems/itis_rrm/utils/DateUtil.java
package za.co.xisystems.itis_rrm.utils;

import android.util.Log;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

/**
 * Created by Mauritz Mollentze on 2014/12/19.
 * Update by Pieter Jacobs during 2016/05, 2016/07.
 */
public class DateUtil {

    // region (Private Static Final Fields)

    private static final String TAG = DateUtil.class.getSimpleName();
    private static final String parsingIso8601DateTimeFailed = "Parsing ISO8601 datetime failed";
    private static final String timeZeros = " 00:00:00";
    private static final String emptyString = "";
    private static final String dash = "-";
    private static final String zero = "0";

    // endregion (Private Static Final Fields)

    // region (Private Static Fields)

    private static DateFormat iso8601Format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
    private static DateFormat readableDateForm = new SimpleDateFormat("dd MMMM yyyy");
    // endregion (Private Static Fields)

    // region (Public Static Methods)

    public static Date StringToDate(String stringDate) {
        if (null == stringDate) return null;

        try {
            return iso8601Format.parse(stringDate);
        } catch (ParseException e) {
            Log.e(TAG, parsingIso8601DateTimeFailed, e);
            return null;
        }
    }

    public static String DateToString(Date date) {
        if (null == date) return null;
        return iso8601Format.format(date);
    }

    public static Date getCurrentDateTime() {
        Date date = new Date();
        return StringToDate(iso8601Format.format(date));
    }

    public static String toStringReadable(Date date) {
        return date == null ? null : readableDateForm.format(date);
    }

    public static String toStringReadable(int year, int month, int day) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(year, month, day);
        return toStringReadable(calendar.getTime());
    }

    public static Date CalendarItemsToDate(int year, int monthOfYear, int dayOfMonth) {
        String mDate;
        String mMon;

        if (dayOfMonth < 10)
            mDate = zero + dayOfMonth;
        else
            mDate = emptyString + dayOfMonth;

        if ((monthOfYear + 1) < 10)
            mMon = zero + (monthOfYear + 1);
        else
            mMon = emptyString + (monthOfYear + 1);

        return StringToDate(emptyString + year + dash + mMon + dash + mDate + timeZeros);
    }

    public static Calendar DateToCalendar(Date date) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        return cal;
    }

    // endregion (Public Static Methods)
}

File: app/src/main/java/za/co/xisystems/itis_rrm/utils/DateUtil.kt
/*
 * Updated by Shaun McDonald on 2021/01/25
 * Last modified on 2021/01/25 6:30 PM
 * Copyright (c) 2021.  XI Systems  - All rights reserved
 */

package za.co.xisystems.itis_rrm.utils

import timber.log.Timber
import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Calendar
import java.util.Date
import java.util.Locale

/**
 * Created by Mauritz Mollentze on 2014/12/19.
 * Update by Pieter Jacobs during 2016/05, 2016/07.
 */
object DateUtil {
    private const val parsingIso8601DateTimeFailed = "Parsing ISO8601 datetime failed"

    private val iso8601Format: DateFormat =
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.ROOT)

    private val localDateTimeFormat: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH)
    private val readableDateForm: DateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)

    fun stringToDate(stringDate: String?): Date? {
        return if (null == stringDate) null else try {
            return if (stringDate.contains('+') || stringDate.lastIndexOf("-") > 8) {
                iso8601Format.parse(stringDate)
            } else {
                val localDateTime = LocalDateTime.parse(stringDate)
                Date.from(localDateTime.atZone(ZoneId.systemDefault()).toInstant())
            }
        } catch (e: ParseException) {
            Timber.e(e, parsingIso8601DateTimeFailed)
            null
        }
    }

    fun dateToLocalString(date: Date?): String? {
        return date?.toInstant()?.toEpochMilli()?.toString()
    }

    fun dateToString(date: Date?): String? {
        return if (null == date) null else iso8601Format.format(date)
    }

    val currentDateTime: Date?
        get() {
            val date = Date()
            return stringToDate(iso8601Format.format(date))
        }

    fun toStringReadable(date: Date?): String? {
        return if (date == null) null else readableDateForm.format(date)
    }

    fun toStringReadable(year: Int, month: Int, day: Int): String? {
        val calendar = Calendar.getInstance()
        calendar[year, month] = day
        return toStringReadable(calendar.time)
    }
}
